libraries/AsyncUDP/src/AsyncUDP.cpp

```diff
@@ -682,9 +698,8 @@ void AsyncUDP::_recv(udp_pcb *upcb, pbuf *pb, const ip_addr_t *addr, uint16_t po
         if(_handler) {
             AsyncUDPPacket packet(this, this_pb, addr, port, netif);
             _handler(packet);
-        } else {
-            pbuf_free(this_pb);
```

Comment:
on destruct the packet will free the pbuf. if it does not, that should be fixed there, or where the packet is copied

Reply:
`pbuf_free`, according to the code, needs to be called at least once, even if the handler isn't used. As soon as the handler is used with the new `AsyncUDPPacket` object, its constructor and destructor will `pbuf_ref` and `pbuf_free`, respectively. That's why the copy constructor needs to add a `pbuf_ref`, to balance things out if the pass-as-copy is used.

What are your thoughts on the logic in #3288? I tested the code with my example from #3287 using both pass-as-reference and pass-as-copy forms.

Reply:
Another possibility for a fix: Don't call `pbuf_ref` in the constructor or copy constructor, and don't call `pbuf_free`in the destructor. Have all management outside the `AsyncUDPPacket` object. However, code that currently exists and assumes the class will manage this will break.